go.mod

```diff
@@ -450,10 +451,11 @@ replace (
 	github.com/mwitkow/go-conntrack => github.com/mwitkow/go-conntrack v0.0.0-20190716064945-2f068394615f
 	github.com/mxk/go-flowrate => github.com/mxk/go-flowrate v0.0.0-20140419014527-cca7078d478f
 	github.com/niemeyer/pretty => github.com/niemeyer/pretty v0.0.0-20200227124842-a10e7caefd8e
-	github.com/nxadm/tail => github.com/nxadm/tail v1.4.4
+	github.com/nxadm/tail => github.com/nxadm/tail v1.4.8
 	github.com/olekukonko/tablewriter => github.com/olekukonko/tablewriter v0.0.4
-	github.com/onsi/ginkgo => github.com/onsi/ginkgo v1.14.0
-	github.com/onsi/gomega => github.com/onsi/gomega v1.10.1
+	github.com/onsi/ginkgo => github.com/onsi/ginkgo v1.16.4
```

Comment:
looks like we still have deps on both github.com/onsi/ginkgo and github.com/onsi/ginkgo/v2?

Reply:
at least for k8s, this is not needed, as the entire directory of `github.com/onsi/ginkgo/ginkgo` has been removed.

Reply:
but its presence here means it's still in our dependency graph

Reply:
> There was a limitation in go until v1.17 where indirect dependencies imported by your dependencies' test packages would get pulled into your go.sum. This led to a terrible dependency loop where Gomega and Ginkgo depended on each other (but only under test) pulling several older versions of Ginkgo/Gomega into a project's go.sum. Thankfully this is [resolved](https://go.dev/doc/go1.17#go-command) beginning in 1.17.

Transitive dependencies still impact our dependency graph... the changes in go1.17 just made pruning of the files nicer.

Reply:
bumped to Ginkgo v2.1.4, and Gomega to  v1.19.0,  still cannot get rid of this dependency, I grepped entire source in k8s and cannot found the Ginkgo V1 is imported anywhere, this should be a indirect dependency and I cannot tell which package has this indirect dependency. Is this something the blocking issue for this pr?

Reply:
@liggitt

Reply:
`go mod graph` can help identify what is pulling it in. In general, we want dependencies to decrease, not increase. We should at least understand what remains to be able to drop the v1 dep and have a concrete plan / timeframe for doing so before taking on the v2 dep.

Reply:
```
$ go mod why -m github.com/onsi/ginkgo
# github.com/onsi/ginkgo
k8s.io/apiextensions-apiserver/pkg/apiserver
k8s.io/kube-openapi/pkg/util/proto
k8s.io/kube-openapi/pkg/util/proto.test
github.com/onsi/ginkgo
```

Here's the culprit:
https://github.com/kubernetes/kube-openapi/blob/5e7f5fdc6da62df0ce329920a63eda22f95b9614/pkg/util/proto/openapi_suite_test.go#L20

As far as I can tell, it's the only reason for the indirect dependency.

Reply:
@apelisse found a ginkgo/v2 → gomega → ginkgo/v1 dependency while reviewing the kube-openapi PR (https://github.com/kubernetes/kube-openapi/pull/302#issuecomment-1159113508)

that should be resolved so ginkgo/v2 doesn't transitively refer to ginkgo/v1

Reply:
`go mod why -m github.com/onsi/ginkgo` in kube-openapi for that PR doesn't report any dependency on ginkgo/v1.

I downloaded all source for it via `GOMODCACHE=/tmp/kube-openapi go mod tidy` and the only reference to ginkgo/v1 is in the go.sum file of gomega. Doing the same for gomega 1.19.0 shows only references ginkgo/v1 in older releases:
```
$ grep -r 'github.com/onsi/ginkgo' /tmp/gomega | grep -v 'ginkgo/v2'
/tmp/gomega/cache/download/github.com/onsi/gomega/@v/v1.7.1.mod:	github.com/onsi/ginkgo v1.6.0
/tmp/gomega/cache/download/github.com/onsi/gomega/@v/v1.10.1.mod:	github.com/onsi/ginkgo v1.12.1
/tmp/gomega/cache/download/github.com/onsi/gomega/@v/v1.17.0.mod:	github.com/onsi/ginkgo v1.16.4
/tmp/gomega/cache/download/github.com/onsi/ginkgo/@v/v1.6.0.mod:module github.com/onsi/ginkgo
/tmp/gomega/cache/download/github.com/onsi/ginkgo/@v/v1.12.1.mod:module github.com/onsi/ginkgo
/tmp/gomega/cache/download/github.com/onsi/ginkgo/@v/v1.16.4.mod:module github.com/onsi/ginkgo
```

I suspect that we can ignore references to ginkgo/v1 in go.sum. I believe it caches also information about versions that aren't needed anymore.

Reply:
`go mod graph` on that PR shows ginkgo/v1 in the graph, so it is definitely more involved than just residual go.sum entries:

```
go mod graph | grep " github.com/onsi/ginkgo@v1"
github.com/onsi/gomega@v1.17.0 github.com/onsi/ginkgo@v1.16.4
github.com/onsi/gomega@v1.10.1 github.com/onsi/ginkgo@v1.12.1
github.com/onsi/gomega@v1.7.1 github.com/onsi/ginkgo@v1.6.0
```

Reply:
`go mod graph` more accurately shows everything considered when resolving module versions

Reply:
how can I help here?  `go mod graph` shows only `ginkgo/v2` when I run it in both the ginkgo and gomega repos with go 1.18.  

note that ginkgo and gomega mutually import each other but only *under test*.  in prior versions of go this test-only dependency resulted in all prior versions getting pulled in.  but as I mentioned [here](https://github.com/kubernetes/kubernetes/pull/109111#discussion_r865978052) go 1.17 introduced a feature that is now the default in go 1.18 (since 1.16 is no longer under support) that effectively avoids pulling in dependencies of test dependencies and breaks the cycle.  I assume y'all are running on go 1.18?

Reply:
ah, maybe it's kube-openapi that needs to bump its `go.mod` to indicate `go1.18` to pick up that improvement

Reply:
verified, the issue will be resolved with https://github.com/kubernetes/kube-openapi/pull/306.

Reply:
Even if it's only used for tests, the chain that @liggitt identified in https://github.com/kubernetes/kubernetes/pull/109111#issuecomment-1159129496 eventually ends up "downloading" github.com/onsi/ginkgo and therefore it remains in go.sum. It's also in gomega v1.19.0: https://github.com/onsi/gomega/blob/3143ff4a92ce52253b4b256cc4c1e8f026c260d5/go.sum#L31

It's not on gomega master. I am not sure whether that is because the ginkgo dependency was updated from v2.1.3 to v2.1.4 in gomega. The difference in ginkgo between those releases may have been significant enough to cut the chain:
```diff
diff --git a/go.mod b/go.mod
index 1703506..4c1fb7d 100644
--- a/go.mod
+++ b/go.mod
@@ -1,11 +1,19 @@
 module github.com/onsi/ginkgo/v2
 
-go 1.16
+go 1.18
 
 require (
        github.com/go-task/slim-sprig v0.0.0-20210107165309-348f09dbbbc0
        github.com/google/pprof v0.0.0-20210407192527-94a9f03dee38
-       github.com/onsi/gomega v1.17.0
-       golang.org/x/sys v0.0.0-20210423082822-04245dca01da
-       golang.org/x/tools v0.0.0-20201224043029-2b0845dc783e
+       github.com/onsi/gomega v1.19.0
+       golang.org/x/sys v0.0.0-20220319134239-a9b59b0215f8
+       golang.org/x/tools v0.1.10
+)
+
+require (
+       github.com/golang/protobuf v1.5.2 // indirect
+       golang.org/x/net v0.0.0-20220225172249-27dd8689420f // indirect
+       golang.org/x/text v0.3.7 // indirect
+       google.golang.org/protobuf v1.26.0 // indirect
+       gopkg.in/yaml.v2 v2.4.0 // indirect
 )
```

@onsi: So perhaps all we need is another gomega release?

Reply:
Never mind. As discussed elsewhere, the relevant change is the switch to `go 1.18` in `go.mod` and we can do the same in our `go.mod` files if we haven't switched already.